src/instructlab/data/list.py

```diff
@@ -0,0 +1,101 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Standard
+from pathlib import Path
+from typing import List, Tuple, TypedDict
+import logging
+import time
+
+# Third Party
+import click
+
+# First Party
+from instructlab import clickext
+from instructlab.configuration import DEFAULTS
+
+logger = logging.getLogger(__name__)
+
+
+class DatasetListing(TypedDict):
+    """
+    Represents a single dataset listing.
+    """
+
+    filename: str
+    created_at: str
+    file_size: str
+
+
+# convert_bytes_to_proper_mag takes a dir/file size in Bytes and converts it to the proper Magnitude (MB, GB)
+def convert_bytes_to_proper_mag(f_size: int) -> Tuple[float, str]:
+    magnitudes = ["KB", "MB", "GB"]
+    magnitude = "B"
+    adjusted_fsize = float(f_size)
+    for mag in magnitudes:
+        if adjusted_fsize >= 1024:
+            magnitude = mag
+            adjusted_fsize /= 1024
+        else:
+            return adjusted_fsize, magnitude
+    return adjusted_fsize, magnitude
+
+
+# print_table displays the datasets on the system in a nicely formatted table
+def print_table(headers: list[str], data: list[list[str]]):
+    column_widths = [
+        max(len(str(row[i])) for row in data + [headers]) for i in range(len(headers))
+    ]
+
+    # Print separator line between headers and data
+    horizontal_lines = ["-" * (width + 2) for width in column_widths]
+    joining_line = "+" + "+".join(horizontal_lines) + "+"
+
+    print(joining_line)
+
+    outputs = []
+    for header, width in zip(headers, column_widths, strict=False):
+        outputs.append(f" {header:{width}} ")
+    print("|" + "|".join(outputs) + "|")
+
+    print(joining_line)
+
+    for row in data:
+        outputs = []
+        for item, width in zip(row, column_widths, strict=False):
+            outputs.append(f" {item:{width}} ")
+        print("|" + "|".join(outputs) + "|")
+
+    print(joining_line)
+
+
+@click.command(name="list")
+@clickext.display_params
+@click.option(
+    "--dataset-dirs",
+    help="Base directories where datasets are stored.",
+    multiple=True,
+    default=[DEFAULTS.DATASETS_DIR],
+    show_default=True,
+)
+def list_data(dataset_dirs: list[str]):
+    """lists datasets"""
+    entries: List[DatasetListing] = []
+    for directory in dataset_dirs:
+        for entry in Path(directory).iterdir():
```

Comment:
This is my ignorance, but what happens if someone passes a mutli-level nested directory? Does `iterdir()` recursively get to the bottom of the tree and only process files? should we be verifying that the entry is of type `.jsonl` here?

Reply:
I think we should filter our anything that is not `isfile`.

Reply:
Will we ever have a case where the SDG dataset is outputted into a directory?

Reply:
Depends on how you'd like to handle a case where something that is not a last level directory was passed. If you are ok with reporting directories in the output (probably with a minuscule size - since we don't drill down like we do in `_analyze_dir` in `model list` implementation), then this is enough.

---

If, as @jaideepr97 seems to suggest, we want to drill down, then `iterdir()` is not the right way. Instead, `os.walk` should be used.

Reply:
I see what you mean. The reason why we should avoid this for the time being is that it introduces confusion in the implementation (rightfully so) which shouldn't be a priority for 0.18.0.
Specifically, by drilling through subdirectories, we would end up printing them in the format of `foo/bar/dataset.jsonl`. This is misleading since it's printed as a relative path, and so the user may end up referencing it as such. However; the default is to look at the user's persisent storage to extract these datasets. In this way, we can at least provide the guidance that the datasets are relative to where they are stored in local storage. Otherwise, it will greatly convulate the work we need to do for documentation and presentation.